rfd/0179-api-versioning.md

```diff
@@ -0,0 +1,392 @@
+}
+```
+ Some users had too many access requests to send over the fetch and the page
+ would break, becoming unusable for larger users. The solution was to server
+ side paginate the Access Requests request. The return used to just be a list of
+ Access Requests, and it has then changed to an object with a list of Access
+ Requests as a field and a nextKey field.
+ 
+ ```go
+// example handler
+func AccessRequestHandler() ([]AccessRequest, error) {
+    accessRequests := []AccessRequest
+    nextKey := "hi_im_next"
+    return AccessRequestResponse{
+        requests: accessRequests,
+        nextKey: nextKey,
+    }, nil
+}
+```
+
+This breaks backward compatibility because a web client of version n-1 would be
+expecting a list of requests only, and a web client of n hitting a proxy of n-1
+would be expecting to receive the paginated list.
+
+The solution _in this particular case_ was
+1. on the proxy side, check which version of the web client was making the request (by the existence of a specific query param, `limit`) and decide which shape to send
+2. on the web client side, if receiving a response from an old proxy, fallback to the client pagination.
+
+This may work, but it highlights a couple things.
+- we can send any response shape we want with no errors/warnings. 
+- if backward compatibility wasn't forefront of the mind, there was nothing stopping us from just pushing the new code with a break
+- the backend has to check the request shape and the frontend has to check the response shape just to know what type it is receiving.
+
+Using a standard guideline for an rpc service should help us evolve our API overtime in a backward compatible way.
+
+## Web Client/API Backward Compatibility Guidelines
+
+In general, if the Request or Response shape does not change, logic in an API handler can change without worry of backward compatibility.
+
+### Creating a new endpoint
+Creating a new endpoint that previously didn't exist is ok. This endpoint should
+be prefixed with `/v1` if using the REST API paradigm. The response shape should
+always return an object, even if that object would contain only one field.
+Prefer pagination for resources that can be paginated, and always create the
+response as if they _could_ be paginated in the future.
+
+Bad
+```go
+func (h *Handler) getFoos() (interface{}, error) {
+    foos := []string{"foo1", "foo2", "foo3"}
+	return foos, nil
+}
+```
+Good
+```go
+type GetFoosResponse struct {
+    Foos []string
+    NextKey string
+    ContainsDuplicates bool
+}
+
+func (h *Handler) getFoos() (interface{}, error) {
+    foos := []string{"foo1", "foo2", "foo3"}
+	return GetFoosResponse{
+	     Foos: foos,
+	     ContainsDuplicates: true
+	}, nil
+}
+```
+
+### Removing an endpoint
+An endpoint can be removed after our backward compatibility promise has been
+fulfilled and NOT before. This means marking for deletion with a TODO and
+removing it in the relevant version. For example, if v17 no longer uses
```

Comment:
I don't think minors count? As far as I can tell, we only worry about [major versions](https://goteleport.com/docs/upgrading/overview/#component-compatibility) when it comes to declaring incompatibility. Why does something in 17.1 have to stay until 19 compared to something in 17.0 being ok to be gone in 18?

Reply:
this is the scenario i was trying to describe: if something got introduced `v17.1`, then you have to mark for deletion in `v19` (not `v18`), b/c if we delete in v18, then we break compatibility with v17.0 -> v18.0

am i misunderstanding? i do get confused by this all the time so maybe i am 😅

Reply:
Oh I see what you are saying. Then yes YOU are correct and I will update the example!

Reply:
oh btw, i think we need to include the `patch` too... probably unlikely to happen but it could happen, so:

- new feature introduce on major version, mark deletion for one major version ahead
- new feature introduce on minor/patch version, mark deletion for two major version ahead

Reply:
Wow, yeah, it didn't occur to me, but yeah, it's something we do have to account for.

Perhaps we could explain the general rule and then show examples, rather than adding one rule and then being like "oh in case a feature is added in a minor or patch version, there's a slightly different rule". I was thinking of something like this:

> An endpoint can be removed in a major version n+2, where n is the last major version where the endpoint was used.
>
> Example 1: v17 no longer uses `GET /webapi/foos` which was last used in v16. The endpoint can be removed in v18.
> Example 2: v4.2.0 no longer uses `GET /webapi/bars` which was last used in v4.1.3. The endpoint can be removed in v6, so that v5 still supports clients using v4.1.3 and v4.2.0.